### PR TITLE
SamFileHeader was not available causing a NPE

### PR DESCRIPTION
**Pull request by nh13**
- [ ] I am running this on a test sample to verify it runs to completion.  I will update this PR once it has succeeded.

### CODE REVIEW

**Comment by yfarjoun** [comment]
I had heard complaints that this code doesn't work. @takutosato here's the bug fix! The fact that this wasn't tested at all is troubling. Could you add a test that just runs the code against a small bam and verifies that it creates a metric?

**Comment by takutosato** [comment]
@yfarjoun will do.

**Comment by takutosato** [comment]
The tools still doesn't run; you get an empty header inside `getIntervalsToExamine()` (the line above super.doWork(). I have a fix though. PR coming in the next day or two.

**Comment by nh13** [comment]
@takutosato I have an incoming fix for that too, so I will append a few tests and fixes to this PR.

**Comment by nh13** [comment]
@takutosato and @yfarjoun I have fixed a number of issues and added some basic tests.  Again, I am going to run this on a large dataset to verify the results of these fixes.

File: src/main/java/picard/analysis/CollectWgsMetrics.java [hunk 1]
```diff
@@ -480,6 +487,7 @@ protected IntervalList getIntervalsToExamine() {
         return intervals;
     }
 
+    /** This method should only be called after {@link this.getSamReader()} is called. */
     protected SAMFileHeader getSamFileHeader() {
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/CollectWgsMetrics.java]: add an exception if not?

File: src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.182149);
+                Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.091075);
+                Assert.assertEquals(metrics.PCT_1X, 1.0);
+            }
+        }
+
+        for (final Histogram<Integer> histogram : output.getAllHistograms()) {
+            if (histogram.getValueLabel().equals("count_WHOLE_GENOME")) {
+                Assert.assertEquals(histogram.get(0).getValue(), 274d);
+            }
```
Comment (yfarjoun) [inline src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java]: } else {

😄

File: src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        output.read(new FileReader(outfile));
+
+        for (final WgsMetricsWithNonZeroCoverage metrics : output.getMetrics()) {
+            if (metrics.CATEGORY == WgsMetricsWithNonZeroCoverage.Category.WHOLE_GENOME) {
+                Assert.assertEquals(metrics.GENOME_TERRITORY, 1210);
+                Assert.assertEquals(metrics.PCT_EXC_MAPQ, 0.271403);
+                Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.182149);
+                Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.091075);
+                Assert.assertEquals(metrics.PCT_1X, 0.107438);
+            }
```
Comment (yfarjoun) [inline src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java]: here

File: src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.182149);
+                Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.091075);
+                Assert.assertEquals(metrics.PCT_1X, 1.0);
+            }
+        }
+
+        for (final Histogram<Integer> histogram : output.getAllHistograms()) {
+            if (histogram.getValueLabel().equals("count_WHOLE_GENOME")) {
+                Assert.assertEquals(histogram.get(0).getValue(), 1080d);
+            }
```
Comment (yfarjoun) [inline src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java]: here

File: src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        output.read(new FileReader(outfile));
+
+        for (final WgsMetricsWithNonZeroCoverage metrics : output.getMetrics()) {
+            if (metrics.CATEGORY == WgsMetricsWithNonZeroCoverage.Category.WHOLE_GENOME) {
+                Assert.assertEquals(metrics.GENOME_TERRITORY, 404);
+                Assert.assertEquals(metrics.PCT_EXC_MAPQ, 0.271403);
+                Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.182149);
+                Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.091075);
+                Assert.assertEquals(metrics.PCT_1X, 0.321782);
+            }
```
Comment (yfarjoun) [inline src/test/java/picard/analysis/CollectWgsMetricsWithNonZeroCoverageTest.java]: here

**Comment by yfarjoun** [comment]
still minor comments, otherwise good-to-go!

**Comment by nh13** [comment]
@yfarjoun I wanted to wait for #658 but on the other hand I'd like to get this in this week (before a release) due to the tool not being useful in its current state.  Given the volume of comments on #658 I'd like to fix this up and merge it.  Thoughts?

**Comment by yfarjoun** [comment]
I think that #658 will still take a while to review and merge...so I think that you can go ahead with fixing this PR. I'd rather have the tool in working condition. The only overlap is the actual solution to the NPE..

**Comment by takutosato** [comment]
Yea, @nh13 please consider my solution to the NPE in #658 (i.e. override `getCollector` in NonZeroCoverage) and choose whichever solution you like. Thanks for giving me a chance to merge first. I wasn't able to deliver but I appreciate it.

**Comment by yfarjoun** [comment]
@nh13 I had made comments on this PR a while ago and it was merged with no reference to my comments... 😕

**Comment by nh13** [comment]
Minor comments fixed so then good to go. See your previous comment. I took that as fix it up then merge.

**Comment by yfarjoun** [comment]
but you didn't fix the comments I raised! the `} else {` comments and also the request that you throw an exception if an assumption isn't met....

**Comment by nh13** [comment]
I see. I can make a quick PR this weekend. What was I thinking?

**Comment by yfarjoun** [comment]
thanks. 👍

**Pull request merged**